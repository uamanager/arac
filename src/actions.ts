export enum Actions {
  CREATE = 'create',
  READ = 'read',
  UPDATE = 'update',
  DELETE = 'delete'
}
