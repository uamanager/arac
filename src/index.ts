export * from './accessControl';
export * from './accessCreate';
export * from './accessRequest';
export * from './actions';
export * from './dynamicChecker';
export * from './permission';
export * from './resource';
export * from './role';
export * from './staticChecker';
