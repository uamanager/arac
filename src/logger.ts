export class Logger {
  constructor (private className: string) {}

  public log(methodName: string, message: string){
    console.log(`Class: '${this.className}', Method: '${methodName}', Logs: '${message}'`);
  }

  public error (methodName: string, message: string) {
    throw new Error(`Class: '${this.className}', Method: '${methodName}', Error: '${message}'`);
  }
}
