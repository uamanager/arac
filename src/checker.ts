import {Actions} from './actions';

export type TCheckerFunction = (
  action: Actions,
  roleName: string,
  resourceName: string
) => Promise<boolean>;

export class Checker {
  constructor (
    public name: string,
    public action: Actions,
    public checker: TCheckerFunction
  ) {
  }

  check (
    action: Actions,
    roleName: string,
    resourceName: string
  ) {
    return this.checker(action, roleName, resourceName);
  }
}
