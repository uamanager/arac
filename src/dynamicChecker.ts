import {Actions} from './actions';

export type TDynamicCheckerFunction = (
  action: Actions,
  roleName: string,
  resourceName: string
) => Promise<boolean>;

export class DynamicChecker {
  constructor (
    public checker: TDynamicCheckerFunction
  ) {
  }

  check (
    action: Actions,
    roleName: string,
    resourceName: string
  ) {
    return this.checker(action, roleName, resourceName);
  }
}
