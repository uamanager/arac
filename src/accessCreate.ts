import {AccessControl} from './accessControl';
import {Actions} from './actions';
import {TDynamicCheckerFunction} from './dynamicChecker';

export class AccessCreate {
  constructor (
    public roleName: string,
    public allowed: boolean,
    public accessControl: AccessControl
  ) {}

  public create (resourceName: string, dynamicCheckers?: {
    [key: string]: TDynamicCheckerFunction
  }) {
    this.accessControl.permission(
      Actions.CREATE,
      AccessControl.hash(this.roleName, resourceName),
      this.allowed,
      dynamicCheckers
    );

    return this;
  }

  public read (resourceName: string, dynamicCheckers?: {
    [key: string]: TDynamicCheckerFunction
  }) {
    this.accessControl.permission(
      Actions.READ,
      AccessControl.hash(this.roleName, resourceName),
      this.allowed,
      dynamicCheckers
    );

    return this;
  }

  public update (resourceName: string, dynamicCheckers?: {
    [key: string]: TDynamicCheckerFunction
  }) {
    this.accessControl.permission(
      Actions.UPDATE,
      AccessControl.hash(this.roleName, resourceName),
      this.allowed,
      dynamicCheckers
    );

    return this;
  }

  public delete (resourceName: string, dynamicCheckers?: {
    [key: string]: TDynamicCheckerFunction
  }) {
    this.accessControl.permission(
      Actions.DELETE,
      AccessControl.hash(this.roleName, resourceName),
      this.allowed,
      dynamicCheckers
    );

    return this;
  }
}
