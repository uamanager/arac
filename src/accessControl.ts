import {AccessCreate} from './accessCreate';
import {AccessRequest} from './accessRequest';
import {Actions} from './actions';
import {TDynamicCheckerFunction} from './dynamicChecker';
import {Logger} from './logger';
import {Permission} from './permission';
import {Resource} from './resource';
import {Role} from './role';

export type TAccessControlDump = {
  role: string,
  resource: string,
  permissions: Partial<{ [key in Actions]: [boolean, ...string[]] }>
}[];

export type TAccessControlDynamicCheckersMap = {
  [key: string]: TDynamicCheckerFunction
};

export class AccessControl {
  private l: Logger = new Logger('AccessControl');

  constructor (
    dump: TAccessControlDump = [],
    checkersMap: TAccessControlDynamicCheckersMap = {}
  ) {
    this.import(dump, checkersMap);
  }

  private _roles: { [key: string]: Role } = {};

  public get roles () {
    return this._roles;
  }

  private _resources: Resource = new Resource('', '', this);

  public get resources () {
    return this._resources;
  }

  private _permissions: { [key: string]: Permission } = {};

  public get permissions () {
    return this._permissions;
  }

  public static hash (roleName: string, resourceName: string): string {
    return [resourceName, roleName].join('#');
  }

  public static unhash (hash: string): string[] {
    return hash.split('#');
  }

  public import (
    dump: TAccessControlDump = [],
    checkersMap: TAccessControlDynamicCheckersMap = {}
  ) {
    dump.forEach((permission) => {
      const hash = AccessControl.hash(permission.role, permission.resource);
      Object.keys(Actions)
        .forEach((actionName) => {
          const action = Actions[actionName];
          if (!permission.permissions.hasOwnProperty(action)) {
            return;
          }
          const [staticChecker, ...dynamicCheckers] = permission.permissions[action];
          const dynamicCheckersFunctions = dynamicCheckers.reduce((
            previousValue,
            functionName: string
          ) => {
            if (checkersMap.hasOwnProperty(functionName)) {
              return {...previousValue, [functionName]: checkersMap[functionName]};
            }
            this.l.error('import', `${functionName} is not defined in checkers map.`);
          }, {});
          this.permission(action, hash, staticChecker, dynamicCheckersFunctions);
          this.l.log('import', `Import finished!`);
        });
    });
  }

  public export (): TAccessControlDump {
    return Object.keys(this.permissions)
      .map(name => {
        let permissions = Object.keys(Actions)
          .reduce((previousValue, actionName) => {
            return {
              ...previousValue,
              [Actions[actionName]]: [
                this.permissions[name].staticChecker.permissions[Actions[actionName]],
                ...Object.keys(this.permissions[name].dynamicCheckers[Actions[actionName]])
              ]
            };
          }, {});

        this.l.log('export', `Export finished!`);
        return {
          role: this.permissions[name].role.name,
          resource: this.permissions[name].resource.name,
          permissions
        };
      });
  }

  public allow (roleName: string): AccessCreate {
    return new AccessCreate(roleName, true, this);
  }

  public deny (roleName: string): AccessCreate {
    return new AccessCreate(roleName, false, this);
  }

  public can (roleName: string): AccessRequest {
    return new AccessRequest(roleName, this);
  }

  public role (roleName: string): Role {
    if (this._roles.hasOwnProperty(roleName)) {
      return this._roles[roleName];
    } else {
      this._roles[roleName] = new Role(roleName, this);
      return this._roles[roleName];
    }
  }

  public resource (resourcePath: string): Resource {
    if (this._resources.check(resourcePath)) {
      return this._resources.get(resourcePath);
    } else {
      return this._resources.add(resourcePath, this);
    }
  }

  public permission (
    action: Actions,
    hash: string,
    patch: boolean,
    dynamicCheckers?: TAccessControlDynamicCheckersMap
  ) {
    if (!this._permissions.hasOwnProperty(hash)) {
      const [resourceName, roleName] = AccessControl.unhash(hash);
      const role = this.role(roleName);
      const resource = this.resource(resourceName);

      this._permissions[hash] = new Permission(role, resource);
    }

    this._permissions[hash].patch({[action]: patch});

    if (dynamicCheckers) {
      Object.keys(dynamicCheckers)
        .forEach(
          name => this._permissions[hash].addChecker(
            action,
            name,
            dynamicCheckers[name]
          ));
    }
  }
}
