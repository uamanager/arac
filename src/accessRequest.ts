import {AccessControl} from './accessControl';
import {Actions} from './actions';
import {Permission} from './permission';

export class AccessRequest {
  constructor (
    public roleName: string,
    public accessControl: AccessControl
  ) {}

  create (resourceName: string) {
    const permissions = this.accessControl.permissions[AccessControl.hash(
      this.roleName,
      resourceName
    )];
    return this.check(Actions.CREATE, resourceName, permissions);
  }

  read (resourceName: string) {
    const permissions = this.accessControl.permissions[AccessControl.hash(
      this.roleName,
      resourceName
    )];
    return this.check(Actions.READ, resourceName, permissions);
  }

  update (resourceName: string) {
    const permissions = this.accessControl.permissions[AccessControl.hash(
      this.roleName,
      resourceName
    )];
    return this.check(Actions.UPDATE, resourceName, permissions);
  }

  delete (resourceName: string) {
    const permissions = this.accessControl.permissions[AccessControl.hash(
      this.roleName,
      resourceName
    )];
    return this.check(Actions.DELETE, resourceName, permissions);
  }

  private check (action: Actions, resourceName: string, permissions: Permission) {
    const dynamicCheckers = permissions.dynamicCheckers[action];
    const dynamicCheckersArray = Object.keys(dynamicCheckers)
      .map(name => dynamicCheckers[name]);
    const checkers = [permissions.staticChecker, ...dynamicCheckersArray];

    return checkers.reduce((prev, next) => {
      return prev.then(() => !!next.check ? next.check(
        action,
        this.roleName,
        resourceName
      ) : Promise.resolve())
        .catch((err) => {
          return Promise.reject(err);
        });
    }, Promise.resolve(true));
  }
}
