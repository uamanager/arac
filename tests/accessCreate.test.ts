import {AccessControl} from '../src';
import {checkersMap} from './data/checkers';

let arac;

beforeEach(() => {
  arac = new AccessControl();
});

afterEach(() => {
  arac = undefined;
});


describe('Access Create CREATE', () => {
  it('creates role, resources and CREATE permission', () => {
    arac.allow('testRole').create('testResource1/testResource2');
    expect(arac).toMatchSnapshot();
  });

  it('creates role, resources and CREATE permission with dynamic checker', () => {
    arac.allow('testRole')
      .create(
        'testResource1/testResource2',
        {dynamicCheckerResolve: checkersMap.dynamicCheckerResolve}
      );
    expect(arac).toMatchSnapshot();
  });
});

describe('Access Create READ', () => {
  it('creates role, resources and READ permission', () => {
    arac.allow('testRole').read('testResource1/testResource2');
    expect(arac).toMatchSnapshot();
  });

  it('creates role, resources and READ permission with dynamic checker', () => {
    arac.allow('testRole')
      .read(
        'testResource1/testResource2',
        {dynamicCheckerReject: checkersMap.dynamicCheckerReject}
      );
    expect(arac).toMatchSnapshot();
  });
});

describe('Access Create UPDATE', () => {
  it('creates role, resources and UPDATE permission', () => {
    arac.allow('testRole').update('testResource1/testResource2');
    expect(arac).toMatchSnapshot();
  });

  it('creates role, resources and UPDATE permission with dynamic checker', () => {
    arac.allow('testRole')
      .update(
        'testResource1/testResource2',
        {dynamicCheckerResolve: checkersMap.dynamicCheckerResolve}
      );
    expect(arac).toMatchSnapshot();
  });
});

describe('Access Create DELETE', () => {
  it('creates role, resources and DELETE permission', () => {
    arac.allow('testRole').delete('testResource1/testResource2');
    expect(arac).toMatchSnapshot();
  });

  it('creates role, resources and DELETE permission with dynamic checker', () => {
    arac.allow('testRole')
      .delete(
        'testResource1/testResource2',
        {dynamicCheckerReject: checkersMap.dynamicCheckerReject}
      );
    expect(arac).toMatchSnapshot();
  });
});
