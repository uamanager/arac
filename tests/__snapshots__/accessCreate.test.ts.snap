// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Access Create CREATE creates role, resources and CREATE permission 1`] = `
AccessControl {
  "_permissions": Object {
    "testResource1/testResource2#testRole": Permission {
      "dynamicCheckers": Object {
        "create": Object {},
        "delete": Object {},
        "read": Object {},
        "update": Object {},
      },
      "resource": Resource {
        "accessControl": [Circular],
        "children": Object {},
        "name": "testResource2",
        "parent": Resource {
          "accessControl": [Circular],
          "children": Object {
            "testResource2": [Circular],
          },
          "name": "testResource1",
          "parent": Resource {
            "accessControl": [Circular],
            "children": Object {
              "testResource1": [Circular],
            },
            "name": "",
            "parent": null,
            "path": "",
          },
          "path": "testResource1",
        },
        "path": "testResource1/testResource2",
      },
      "role": Role {
        "accessControl": [Circular],
        "name": "testRole",
      },
      "staticChecker": StaticChecker {
        "permissions": Object {
          "create": true,
          "delete": false,
          "read": false,
          "update": false,
        },
      },
    },
  },
  "_resources": Resource {
    "accessControl": [Circular],
    "children": Object {
      "testResource1": Resource {
        "accessControl": [Circular],
        "children": Object {
          "testResource2": Resource {
            "accessControl": [Circular],
            "children": Object {},
            "name": "testResource2",
            "parent": [Circular],
            "path": "testResource1/testResource2",
          },
        },
        "name": "testResource1",
        "parent": [Circular],
        "path": "testResource1",
      },
    },
    "name": "",
    "parent": null,
    "path": "",
  },
  "_roles": Object {
    "testRole": Role {
      "accessControl": [Circular],
      "name": "testRole",
    },
  },
  "l": Logger {
    "className": "AccessControl",
  },
}
`;

exports[`Access Create CREATE creates role, resources and CREATE permission with dynamic checker 1`] = `
AccessControl {
  "_permissions": Object {
    "testResource1/testResource2#testRole": Permission {
      "dynamicCheckers": Object {
        "create": Object {
          "dynamicCheckerResolve": DynamicChecker {
            "checker": [MockFunction],
          },
        },
        "delete": Object {},
        "read": Object {},
        "update": Object {},
      },
      "resource": Resource {
        "accessControl": [Circular],
        "children": Object {},
        "name": "testResource2",
        "parent": Resource {
          "accessControl": [Circular],
          "children": Object {
            "testResource2": [Circular],
          },
          "name": "testResource1",
          "parent": Resource {
            "accessControl": [Circular],
            "children": Object {
              "testResource1": [Circular],
            },
            "name": "",
            "parent": null,
            "path": "",
          },
          "path": "testResource1",
        },
        "path": "testResource1/testResource2",
      },
      "role": Role {
        "accessControl": [Circular],
        "name": "testRole",
      },
      "staticChecker": StaticChecker {
        "permissions": Object {
          "create": true,
          "delete": false,
          "read": false,
          "update": false,
        },
      },
    },
  },
  "_resources": Resource {
    "accessControl": [Circular],
    "children": Object {
      "testResource1": Resource {
        "accessControl": [Circular],
        "children": Object {
          "testResource2": Resource {
            "accessControl": [Circular],
            "children": Object {},
            "name": "testResource2",
            "parent": [Circular],
            "path": "testResource1/testResource2",
          },
        },
        "name": "testResource1",
        "parent": [Circular],
        "path": "testResource1",
      },
    },
    "name": "",
    "parent": null,
    "path": "",
  },
  "_roles": Object {
    "testRole": Role {
      "accessControl": [Circular],
      "name": "testRole",
    },
  },
  "l": Logger {
    "className": "AccessControl",
  },
}
`;

exports[`Access Create DELETE creates role, resources and DELETE permission 1`] = `
AccessControl {
  "_permissions": Object {
    "testResource1/testResource2#testRole": Permission {
      "dynamicCheckers": Object {
        "create": Object {},
        "delete": Object {},
        "read": Object {},
        "update": Object {},
      },
      "resource": Resource {
        "accessControl": [Circular],
        "children": Object {},
        "name": "testResource2",
        "parent": Resource {
          "accessControl": [Circular],
          "children": Object {
            "testResource2": [Circular],
          },
          "name": "testResource1",
          "parent": Resource {
            "accessControl": [Circular],
            "children": Object {
              "testResource1": [Circular],
            },
            "name": "",
            "parent": null,
            "path": "",
          },
          "path": "testResource1",
        },
        "path": "testResource1/testResource2",
      },
      "role": Role {
        "accessControl": [Circular],
        "name": "testRole",
      },
      "staticChecker": StaticChecker {
        "permissions": Object {
          "create": false,
          "delete": true,
          "read": false,
          "update": false,
        },
      },
    },
  },
  "_resources": Resource {
    "accessControl": [Circular],
    "children": Object {
      "testResource1": Resource {
        "accessControl": [Circular],
        "children": Object {
          "testResource2": Resource {
            "accessControl": [Circular],
            "children": Object {},
            "name": "testResource2",
            "parent": [Circular],
            "path": "testResource1/testResource2",
          },
        },
        "name": "testResource1",
        "parent": [Circular],
        "path": "testResource1",
      },
    },
    "name": "",
    "parent": null,
    "path": "",
  },
  "_roles": Object {
    "testRole": Role {
      "accessControl": [Circular],
      "name": "testRole",
    },
  },
  "l": Logger {
    "className": "AccessControl",
  },
}
`;

exports[`Access Create DELETE creates role, resources and DELETE permission with dynamic checker 1`] = `
AccessControl {
  "_permissions": Object {
    "testResource1/testResource2#testRole": Permission {
      "dynamicCheckers": Object {
        "create": Object {},
        "delete": Object {
          "dynamicCheckerReject": DynamicChecker {
            "checker": [MockFunction],
          },
        },
        "read": Object {},
        "update": Object {},
      },
      "resource": Resource {
        "accessControl": [Circular],
        "children": Object {},
        "name": "testResource2",
        "parent": Resource {
          "accessControl": [Circular],
          "children": Object {
            "testResource2": [Circular],
          },
          "name": "testResource1",
          "parent": Resource {
            "accessControl": [Circular],
            "children": Object {
              "testResource1": [Circular],
            },
            "name": "",
            "parent": null,
            "path": "",
          },
          "path": "testResource1",
        },
        "path": "testResource1/testResource2",
      },
      "role": Role {
        "accessControl": [Circular],
        "name": "testRole",
      },
      "staticChecker": StaticChecker {
        "permissions": Object {
          "create": false,
          "delete": true,
          "read": false,
          "update": false,
        },
      },
    },
  },
  "_resources": Resource {
    "accessControl": [Circular],
    "children": Object {
      "testResource1": Resource {
        "accessControl": [Circular],
        "children": Object {
          "testResource2": Resource {
            "accessControl": [Circular],
            "children": Object {},
            "name": "testResource2",
            "parent": [Circular],
            "path": "testResource1/testResource2",
          },
        },
        "name": "testResource1",
        "parent": [Circular],
        "path": "testResource1",
      },
    },
    "name": "",
    "parent": null,
    "path": "",
  },
  "_roles": Object {
    "testRole": Role {
      "accessControl": [Circular],
      "name": "testRole",
    },
  },
  "l": Logger {
    "className": "AccessControl",
  },
}
`;

exports[`Access Create READ creates role, resources and READ permission 1`] = `
AccessControl {
  "_permissions": Object {
    "testResource1/testResource2#testRole": Permission {
      "dynamicCheckers": Object {
        "create": Object {},
        "delete": Object {},
        "read": Object {},
        "update": Object {},
      },
      "resource": Resource {
        "accessControl": [Circular],
        "children": Object {},
        "name": "testResource2",
        "parent": Resource {
          "accessControl": [Circular],
          "children": Object {
            "testResource2": [Circular],
          },
          "name": "testResource1",
          "parent": Resource {
            "accessControl": [Circular],
            "children": Object {
              "testResource1": [Circular],
            },
            "name": "",
            "parent": null,
            "path": "",
          },
          "path": "testResource1",
        },
        "path": "testResource1/testResource2",
      },
      "role": Role {
        "accessControl": [Circular],
        "name": "testRole",
      },
      "staticChecker": StaticChecker {
        "permissions": Object {
          "create": false,
          "delete": false,
          "read": true,
          "update": false,
        },
      },
    },
  },
  "_resources": Resource {
    "accessControl": [Circular],
    "children": Object {
      "testResource1": Resource {
        "accessControl": [Circular],
        "children": Object {
          "testResource2": Resource {
            "accessControl": [Circular],
            "children": Object {},
            "name": "testResource2",
            "parent": [Circular],
            "path": "testResource1/testResource2",
          },
        },
        "name": "testResource1",
        "parent": [Circular],
        "path": "testResource1",
      },
    },
    "name": "",
    "parent": null,
    "path": "",
  },
  "_roles": Object {
    "testRole": Role {
      "accessControl": [Circular],
      "name": "testRole",
    },
  },
  "l": Logger {
    "className": "AccessControl",
  },
}
`;

exports[`Access Create READ creates role, resources and READ permission with dynamic checker 1`] = `
AccessControl {
  "_permissions": Object {
    "testResource1/testResource2#testRole": Permission {
      "dynamicCheckers": Object {
        "create": Object {},
        "delete": Object {},
        "read": Object {
          "dynamicCheckerReject": DynamicChecker {
            "checker": [MockFunction],
          },
        },
        "update": Object {},
      },
      "resource": Resource {
        "accessControl": [Circular],
        "children": Object {},
        "name": "testResource2",
        "parent": Resource {
          "accessControl": [Circular],
          "children": Object {
            "testResource2": [Circular],
          },
          "name": "testResource1",
          "parent": Resource {
            "accessControl": [Circular],
            "children": Object {
              "testResource1": [Circular],
            },
            "name": "",
            "parent": null,
            "path": "",
          },
          "path": "testResource1",
        },
        "path": "testResource1/testResource2",
      },
      "role": Role {
        "accessControl": [Circular],
        "name": "testRole",
      },
      "staticChecker": StaticChecker {
        "permissions": Object {
          "create": false,
          "delete": false,
          "read": true,
          "update": false,
        },
      },
    },
  },
  "_resources": Resource {
    "accessControl": [Circular],
    "children": Object {
      "testResource1": Resource {
        "accessControl": [Circular],
        "children": Object {
          "testResource2": Resource {
            "accessControl": [Circular],
            "children": Object {},
            "name": "testResource2",
            "parent": [Circular],
            "path": "testResource1/testResource2",
          },
        },
        "name": "testResource1",
        "parent": [Circular],
        "path": "testResource1",
      },
    },
    "name": "",
    "parent": null,
    "path": "",
  },
  "_roles": Object {
    "testRole": Role {
      "accessControl": [Circular],
      "name": "testRole",
    },
  },
  "l": Logger {
    "className": "AccessControl",
  },
}
`;

exports[`Access Create UPDATE creates role, resources and UPDATE permission 1`] = `
AccessControl {
  "_permissions": Object {
    "testResource1/testResource2#testRole": Permission {
      "dynamicCheckers": Object {
        "create": Object {},
        "delete": Object {},
        "read": Object {},
        "update": Object {},
      },
      "resource": Resource {
        "accessControl": [Circular],
        "children": Object {},
        "name": "testResource2",
        "parent": Resource {
          "accessControl": [Circular],
          "children": Object {
            "testResource2": [Circular],
          },
          "name": "testResource1",
          "parent": Resource {
            "accessControl": [Circular],
            "children": Object {
              "testResource1": [Circular],
            },
            "name": "",
            "parent": null,
            "path": "",
          },
          "path": "testResource1",
        },
        "path": "testResource1/testResource2",
      },
      "role": Role {
        "accessControl": [Circular],
        "name": "testRole",
      },
      "staticChecker": StaticChecker {
        "permissions": Object {
          "create": false,
          "delete": false,
          "read": false,
          "update": true,
        },
      },
    },
  },
  "_resources": Resource {
    "accessControl": [Circular],
    "children": Object {
      "testResource1": Resource {
        "accessControl": [Circular],
        "children": Object {
          "testResource2": Resource {
            "accessControl": [Circular],
            "children": Object {},
            "name": "testResource2",
            "parent": [Circular],
            "path": "testResource1/testResource2",
          },
        },
        "name": "testResource1",
        "parent": [Circular],
        "path": "testResource1",
      },
    },
    "name": "",
    "parent": null,
    "path": "",
  },
  "_roles": Object {
    "testRole": Role {
      "accessControl": [Circular],
      "name": "testRole",
    },
  },
  "l": Logger {
    "className": "AccessControl",
  },
}
`;

exports[`Access Create UPDATE creates role, resources and UPDATE permission with dynamic checker 1`] = `
AccessControl {
  "_permissions": Object {
    "testResource1/testResource2#testRole": Permission {
      "dynamicCheckers": Object {
        "create": Object {},
        "delete": Object {},
        "read": Object {},
        "update": Object {
          "dynamicCheckerResolve": DynamicChecker {
            "checker": [MockFunction],
          },
        },
      },
      "resource": Resource {
        "accessControl": [Circular],
        "children": Object {},
        "name": "testResource2",
        "parent": Resource {
          "accessControl": [Circular],
          "children": Object {
            "testResource2": [Circular],
          },
          "name": "testResource1",
          "parent": Resource {
            "accessControl": [Circular],
            "children": Object {
              "testResource1": [Circular],
            },
            "name": "",
            "parent": null,
            "path": "",
          },
          "path": "testResource1",
        },
        "path": "testResource1/testResource2",
      },
      "role": Role {
        "accessControl": [Circular],
        "name": "testRole",
      },
      "staticChecker": StaticChecker {
        "permissions": Object {
          "create": false,
          "delete": false,
          "read": false,
          "update": true,
        },
      },
    },
  },
  "_resources": Resource {
    "accessControl": [Circular],
    "children": Object {
      "testResource1": Resource {
        "accessControl": [Circular],
        "children": Object {
          "testResource2": Resource {
            "accessControl": [Circular],
            "children": Object {},
            "name": "testResource2",
            "parent": [Circular],
            "path": "testResource1/testResource2",
          },
        },
        "name": "testResource1",
        "parent": [Circular],
        "path": "testResource1",
      },
    },
    "name": "",
    "parent": null,
    "path": "",
  },
  "_roles": Object {
    "testRole": Role {
      "accessControl": [Circular],
      "name": "testRole",
    },
  },
  "l": Logger {
    "className": "AccessControl",
  },
}
`;
